api/build/graph.go

```diff
@@ -0,0 +1,594 @@
+	canceled   int
+	skipped    int
+	errored    int
+	startedAt  int
+	finishedAt int
+}
+
+const (
+	// clusters determine graph orientation.
+	BuiltInCluster       = 2
+	PipelineCluster      = 1
+	ServiceCluster       = 0
+	GraphComplexityLimit = 1000 // arbitrary value to limit render complexity.
+)
+
+// swagger:operation GET /api/v1/repos/{org}/{repo}/builds/{build}/graph builds GetBuildGraph
+//
+// Get directed a-cyclical graph for a build in the configured backend
+//
+// ---
+// produces:
+// - application/json
+// parameters:
+// - in: path
+//   name: org
+//   description: Name of the org
+//   required: true
+//   type: string
+// - in: path
+//   name: repo
+//   description: Name of the repo
+//   required: true
+//   type: string
+// - in: path
+//   name: build
+//   description: Build number
+//   required: true
+//   type: integer
+// security:
+//   - ApiKeyAuth: []
+// responses:
+//   '200':
+//     description: Successfully retrieved graph for the build
+//     type: json
+//     schema:
+//       "$ref": "#/definitions/Graph"
+//   '401':
+//     description: Unable to retrieve graph for the build — unauthorized
+//     schema:
+//       "$ref": "#/definitions/Error"
+//   '404':
+//     description: Unable to retrieve graph for the build — not found
+//     schema:
+//       "$ref": "#/definitions/Error"
+//   '500':
+//     description: Unable to retrieve graph for the build
+//     schema:
+//       "$ref": "#/definitions/Error"
+
+// GetBuildGraph represents the API handler to capture a
+// directed a-cyclical graph for a build from the configured backend.
+//
+//nolint:funlen,goconst // ignore function length and constants
+func GetBuildGraph(c *gin.Context) {
+	// capture middleware values
+	b := build.Retrieve(c)
+	o := org.Retrieve(c)
+	r := repo.Retrieve(c)
+	u := user.Retrieve(c)
+	m := c.MustGet("metadata").(*types.Metadata)
+	ctx := c.Request.Context()
+
+	// update engine logger with API metadata
+	//
+	// https://pkg.go.dev/github.com/sirupsen/logrus?tab=doc#Entry.WithFields
+	entry := fmt.Sprintf("%s/%d", r.GetFullName(), b.GetNumber())
+	logger := logrus.WithFields(logrus.Fields{
+		"build": b.GetNumber(),
+		"org":   o,
+		"repo":  r.GetName(),
+		"user":  u.GetName(),
+	})
+
+	baseErr := "unable to retrieve graph"
+
+	logger.Infof("constructing graph for build %s", entry)
+
+	// retrieve the steps for the build from the step table
+	steps := []*library.Step{}
+	page := 1
+	perPage := 100
+
+	for page > 0 {
+		// retrieve build steps (per page) from the database
+		stepsPart, stepsCount, err := database.FromContext(c).ListStepsForBuild(b, nil, page, perPage)
+		if err != nil {
+			retErr := fmt.Errorf("unable to retrieve steps for build %s: %w", entry, err)
+
+			util.HandleError(c, http.StatusNotFound, retErr)
+
+			return
+		}
+
+		// add page of steps to list steps
+		steps = append(steps, stepsPart...)
+
+		// assume no more pages exist if under 100 results are returned
+		if stepsCount < 100 {
```

Comment:
should we use `perPage` instead of `100` here? same for services.

Reply:
good call, easy!